File: src/feed/store.py
from pathlib import Path

import yaml

from src.feed.models import DuplicateFeedError, Feed


class YamlFeedStore:
    def __init__(self, path: str | Path):
        self._path = Path(path)

    def load_feeds(self) -> list[Feed]:
        if not self._path.exists():
            return []
        content = self._path.read_text()
        if not content.strip():
            return []
        data = yaml.safe_load(content)
        if not data or "feeds" not in data:
            return []
        return [Feed(name=f["name"], url=f["url"]) for f in data["feeds"]]

    def save_feeds(self, feeds: list[Feed]) -> None:
        self._path.parent.mkdir(parents=True, exist_ok=True)
        data = {"feeds": [{"name": f.name, "url": f.url} for f in feeds]}
        self._path.write_text(yaml.dump(data, default_flow_style=False))

    def add_feed(self, feed: Feed) -> None:
        feeds = self.load_feeds()
        for existing in feeds:
            if existing.url == feed.url:
                raise DuplicateFeedError(f"Feed with URL '{feed.url}' already exists")
        feeds.append(feed)
        self.save_feeds(feeds)

    def ensure_data_file(self) -> None:
        if not self._path.exists():
            self._path.parent.mkdir(parents=True, exist_ok=True)
            self._path.write_text(yaml.dump({"feeds": []}, default_flow_style=False))

File: _imported/lexora-feed/feed/store.go
package feed

import (
	"errors"
	"os"

	"gopkg.in/yaml.v3"
)

var ErrDuplicateFeed = errors.New("feed URL already exists")

type Feed struct {
	Name string `yaml:"name" json:"name"`
	URL  string `yaml:"url" json:"url"`
}

type feedsFile struct {
	Feeds []Feed `yaml:"feeds"`
}

func LoadFeeds(path string) ([]Feed, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	if len(data) == 0 {
		return nil, nil
	}

	var ff feedsFile
	if err := yaml.Unmarshal(data, &ff); err != nil {
		return nil, err
	}
	return ff.Feeds, nil
}

func SaveFeeds(path string, feeds []Feed) error {
	ff := feedsFile{Feeds: feeds}
	data, err := yaml.Marshal(&ff)
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

func AddFeed(path string, feed Feed) error {
	existing, err := LoadFeeds(path)
	if err != nil {
		return err
	}
	for _, f := range existing {
		if f.URL == feed.URL {
			return ErrDuplicateFeed
		}
	}
	existing = append(existing, feed)
	return SaveFeeds(path, existing)
}

func EnsureDataFile(path string) error {
	// ensure the file exists and can be read
	b, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			_, err := os.Create(path)
			return err
		}
		return err
	}

	// verify write permission is granted
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	_, err = f.Write(b)
	if err != nil {
		return err
	}
	return nil
}
